File: app/src/main/java/com/example/milestonesapplication/View/InfoWindow.java
package com.example.milestonesapplication.View;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;

import com.example.milestonesapplication.R;
import com.example.milestonesapplication.databinding.WindowInfoLayoutBinding;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

public class InfoWindow extends DialogFragment implements OnMapReadyCallback {

    public static final String TAG = InfoWindow.class.getName();

    private WindowInfoLayoutBinding binding;
    private GoogleMap googleMap;

    private String name;
    private String description;
    private LatLng location;

    private static final String ARG_NAME = "name";
    private static final String ARG_DESCRIPTION = "description";
    private static final String ARG_LOCATION = "location";

    public static InfoWindow newInstance(String title,
                                         String description,
                                         LatLng location) {
        InfoWindow fragment = new InfoWindow();
        Bundle args = new Bundle();
        args.putString(ARG_NAME, title);
        args.putString(ARG_DESCRIPTION, description);
        args.putParcelable(ARG_LOCATION, location);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null) {
            name = args.getString(ARG_NAME);
            description = args.getString(ARG_DESCRIPTION);
            location = args.getParcelable(ARG_LOCATION);
        }

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.window_info_layout, container, false);
        binding.name.setText(name);
        binding.description.setText(description);
        String locationString = location.latitude + " " + location.longitude;
        binding.location.setText(locationString);
        binding.backButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
            }
        });

        SupportMapFragment mapFragment = null;
        if (getFragmentManager() != null) {
            mapFragment = (SupportMapFragment) getFragmentManager()
                    .findFragmentById(R.id.map_info_window);
        }
        if (mapFragment == null) {
            return null;
        }
        mapFragment.getMapAsync(this);

        return binding.getRoot();
    }

    @Override
    public void onStart() {
        super.onStart();
        Dialog dialog = getDialog();
        if (dialog != null) {
            int width = ViewGroup.LayoutParams.MATCH_PARENT;
            int height = ViewGroup.LayoutParams.MATCH_PARENT;
            if (dialog.getWindow() != null) {
                dialog.getWindow().setLayout(width, height);
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Fragment f = null;
        if (getFragmentManager() != null) {
            f = getFragmentManager().findFragmentById(R.id.map_info_window);
        }
        if (f != null) {
            getFragmentManager().beginTransaction().remove(f).commit();
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
        setGoogleMap();
    }

    private void setGoogleMap() {
        googleMap.addMarker(new MarkerOptions()
                .position(location));
        googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(location, 15));
        this.googleMap.getUiSettings().setZoomControlsEnabled(false);
        this.googleMap.getUiSettings().setZoomGesturesEnabled(false);
        this.googleMap.getUiSettings().setScrollGesturesEnabled(false);
        this.googleMap.getUiSettings().setRotateGesturesEnabled(false);
        this.googleMap.getUiSettings().setMapToolbarEnabled(false);
    }
}

File: app/src/main/java/com/example/milestonesapplication/views/InfoWindow.kt
package com.example.milestonesapplication.views

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import com.example.milestonesapplication.R
import com.example.milestonesapplication.databinding.WindowInfoLayoutBinding
import com.example.milestonesapplication.models.Milestone
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions

private const val ARG_NAME = "name"
private const val ARG_DESCRIPTION = "description"
private const val ARG_LOCATION = "location"

class InfoWindow : DialogFragment(), OnMapReadyCallback {

    private lateinit var binding: WindowInfoLayoutBinding
    private lateinit var googleMap: GoogleMap

    private var name: String? = null
    private var description: String? = null
    private var location: LatLng? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            name = it.getString(ARG_NAME)
            description = it.getString(ARG_DESCRIPTION)
            location = it.getParcelable(ARG_LOCATION)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = WindowInfoLayoutBinding.inflate(inflater, container, false)
        binding.milestone = Milestone(name!!, description!!, location!!)
        binding.parentView = this

        val mapFragment: SupportMapFragment =
                fragmentManager?.findFragmentById(R.id.map_info_window) as SupportMapFragment?
                ?: return null
        mapFragment.getMapAsync(this)

        return binding.root
    }

    override fun onStart() {
        super.onStart()
        dialog?.window?.setLayout(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        val f = fragmentManager?.findFragmentById(R.id.map_info_window)
        fragmentManager?.beginTransaction()?.remove(f!!)?.commit()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        setGoogleMap()
    }

    private fun setGoogleMap() {
        googleMap.apply {
            addMarker(MarkerOptions().position(location!!))
            moveCamera(CameraUpdateFactory.newLatLngZoom(location, 15f))
            uiSettings.isZoomControlsEnabled = false
            uiSettings.isZoomGesturesEnabled = false
            uiSettings.isScrollGesturesEnabled = false
            uiSettings.isRotateGesturesEnabled = false
            uiSettings.isMapToolbarEnabled = false
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(name: String, description: String, location: LatLng): InfoWindow {
            val fragment = InfoWindow()
            val args = Bundle()
            args.putString(ARG_NAME, name)
            args.putString(ARG_DESCRIPTION, description)
            args.putParcelable(ARG_LOCATION, location)
            fragment.arguments = args
            return fragment
        }

        @JvmField
        val TAG = InfoWindow::class.java.name
    }
}
